python/ray/util/client/dataclient.py

```diff
@@ -101,54 +126,111 @@ def _process_response(self, response: Any) -> None:
+                if response.req_id in self.outstanding_requests:
+                    del self.outstanding_requests[response.req_id]
+                    # Acknowledge response
+                    self._acknowledge(response.req_id)
         else:
             with self.lock:
                 self.ready_data[response.req_id] = response
                 self.cv.notify_all()
 
-    def _process_rpc_error(self, e: grpc.RpcError):
+    def _can_reconnect(self, e: grpc.RpcError) -> bool:
         """
         Processes RPC errors that occur while reading from data stream.
+        Returns True if the error can be recovered from, False otherwise.
         """
-        self._shutdown(e)
-
-        if e.code() == grpc.StatusCode.CANCELLED:
-            # Gracefully shutting down
-            logger.info("Cancelling data channel")
-        elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                          grpc.StatusCode.RESOURCE_EXHAUSTED):
-            # TODO(barakmich): The server may have
-            # dropped. In theory, we can retry, as per
-            # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-            # in practice we may need to think about the correct semantics
-            # here.
-            logger.info("Server disconnected from data channel")
-        else:
-            logger.exception("Got Error from data channel -- shutting down:")
-
-    def _shutdown(self, e: grpc.RpcError) -> None:
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            logger.debug(e)
+            return False
+        logger.debug("Recoverable error in data channel.")
+        logger.debug(e)
+        return True
+
+    def _shutdown(self) -> None:
         """
         Shutdown the data channel
         """
         with self.lock:
             self._in_shutdown = True
-            self._last_exception = e
             self.cv.notify_all()
 
             callbacks = self.asyncio_waiting_data.values()
             self.asyncio_waiting_data = {}
 
-        # Abort async requests with the error.
-        err = ConnectionError("Failed during this or a previous request. "
-                              f"Exception that broke the connection: {e}")
+        if self._last_exception:
+            # Abort async requests with the error.
+            err = ConnectionError(
+                "Failed during this or a previous request. Exception that "
+                f"broke the connection: {self._last_exception}")
+        else:
+            err = ConnectionError(
+                "Request cannot be fulfilled because the data client has "
+                "disconnected.")
         for callback in callbacks:
-            callback(err)
+            if callback:
+                callback(err)
         # Since self._in_shutdown is set to True, no new item
         # will be added to self.asyncio_waiting_data
 
+    def _acknowledge(self, req_id: int) -> None:
+        """
```

Comment:
How is the performance impact here? Does this mean that we are doubling the traffic from cli -> server?

Reply:
Another question: why do we need ack here? shouldn't it be handled by grpc already? For example, when grpc send request and get successfully ack, it means it's received right?

Reply:
At the moment the ack's are sent on every 10th response received, so closer to 10% more requests (and 10% or less more traffic, since Ack's are the smallest message size for data requests). IIUC this should be pretty negligible, since the stream is always open there's less overhead per message than a unary call, and since we don't wait on a response it shouldn't block later requests.

These are application level ack's that the client uses to signal to the server when it can clean responses in its cache. Afaik there's no mechanism to check on the server side when a response is received using gRPC's underlying ack mechanism.